src/openMVG/robust_estimation/rand_sampling.hpp

```diff
@@ -25,85 +25,156 @@
+  static_assert(std::is_integral<IntT>::value, "Only integer types are supported");
+
+  
   std::random_device rd;
-  std::default_random_engine e1(rd());
-  std::uniform_int_distribution<IntT> uniform_dist(0, total_samples-1);
-  while (samples->size() < num_samples)
+  std::mt19937 generator(rd());
+
+  if(numSamples * 1.5 > rangeSize)
+  {
+    // if the number of required samples is a large fraction of the range size
+    // generate a vector with all the elements in the range, shuffle it and 
+    // return the first numSample elements.
+    // this should be more time efficient than drawing at each time.
+    std::vector<IntT> result(rangeSize);
+    std::iota(result.begin(), result.end(), lowerBound);
+    std::shuffle(result.begin(), result.end(), generator);
+    result.resize(numSamples);
+    return result;
+  }
+  else
   {
-    IntT sample = uniform_dist(e1);
-    samples->insert(sample);
+    // otherwise if the number of required samples is small wrt the range
+    // use the optimized Robert Floyd algorithm.
+    // this has linear complexity and minimize the memory usage.
+    std::unordered_set<IntT> samples;
+    for(IntT d = upperBound - numSamples; d < upperBound; ++d)
+    {
+      IntT t = std::uniform_int_distribution<>(0, d)(generator) + lowerBound;
+      if(samples.find(t) == samples.end())
+        samples.insert(t);
+      else
+        samples.insert(d);
+    }
+    assert(samples.size() == numSamples);
+    std::vector<IntT> result(std::make_move_iterator(samples.begin()),
+                             std::make_move_iterator(samples.end()));
+    return result;
   }
 }
 
+/**
+* @brief Pick a random subset of the integers in the range [0, upperBound).
+*
+* @param[in] numSamples The number of samples to produce.
+* @param[in] upperBound The upper bound of the range.
+* @param[out] samples The set containing the random numbers in the range [0, upperBound)
+*/
 template<typename IntT>
-static void UniformSample(
-  size_t num_samples,
-  size_t total_samples,
-  std::vector<IntT> *samples)
+inline void UniformSample(std::size_t numSamples,
+                          std::size_t upperBound,
+                          std::set<IntT> &samples)
 {
-  assert(num_samples <= total_samples);
-  samples->resize(0);
-  samples->reserve(num_samples);
-  std::random_device rd;
-  std::default_random_engine e1(rd());
-  std::uniform_int_distribution<IntT> uniform_dist(0, total_samples-1);
-  std::set<IntT> set_samples;
-  while (set_samples.size() < num_samples)
+  assert(numSamples <= upperBound);
+  static_assert(std::is_integral<IntT>::value, "Only integer types are supported");
+  
+  const auto vecSamples = randSample<IntT>(0, upperBound, numSamples);
+  for(const auto& s : vecSamples)
   {
-    IntT sample = uniform_dist(e1);
-    if(set_samples.count(sample) == 0)
-    {
-      set_samples.insert(sample);
-      samples->push_back(sample);
-    }
+    samples.insert(s);
   }
+  assert(samples.size() == numSamples);
```

Comment:
same as above, it is `==`, this is a wrapper around the same function in order to get a `std::set` as output instead of a `std::vector`, because somewhere in the lib there is a case in which the samples must be ordered. It could be also inside the if
```
assert(samples.insert(s).second)
```
as the vector vecSamples must have only unique elements